### PR TITLE
PERF: reindex with MultiIndex

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.0.rst` file if fixing a bug or adding a new feature.

Avoid unnecessary (and potentially expensive) call to MultiIndex._values. This only matters when MultiIndex._values has not been cached. Existing asv has been updated to test cached and non-cached cases.

```
import numpy as np
import pandas as pd

mi = pd.MultiIndex.from_arrays(
  [np.arange(10**6)] * 2
)
mi2 = pd.MultiIndex.from_arrays(
  [np.arange(10**7)] * 2
)

df = pd.DataFrame({"A": 1.0}, index=mi)

%timeit x = df.reindex(mi2.copy())
```

```
1.84 s ± 106 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)   <- main
685 ms ± 22.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  <- PR
```

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3922,14 +3922,15 @@ def _get_indexer(
         elif method == "nearest":
             indexer = self._get_nearest_indexer(target, limit, tolerance)
         else:
-            tgt_values = target._get_engine_target()
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:3925]: nice!

**Comment by jreback** [comment]
very nice @lukemanley

**Comment by twoertwein** [comment]
I think this PR made pyright unhappy, but I honestly don't see how this PR affected _mixins.py:

> pandas/core/arrays/_mixins.py:185:26 - warning: Instance methods should take a "self" parameter (reportSelfClsParameterName)

If the solution is to ignore this warning, you can add
`# pyright: reportSelfClsParameterName = false`
at the top of the file.

**Comment by jreback** [comment]
i think was a prior PR

**Comment by twoertwein** [comment]
> i think was a prior PR

Yes, you are right: #46214 If no one beats me, I will create a PR to rename cls to self

**Pull request merged**